File: examples/src/Router.fs
[<RequireQualifiedAccess>]
module Router

open Elmish.UrlParser
open Elmish.Navigation
open Feliz
open Browser.Dom

[<RequireQualifiedAccess>]
type ComposabilityRoute =
    | Simple
    | WithConfiguration

module ComposabilityRoute =

    let toSegment (route : ComposabilityRoute) =
        match route with
        | ComposabilityRoute.Simple -> "simple"
        | ComposabilityRoute.WithConfiguration -> "with-configuration"

[<RequireQualifiedAccess>]
type Route =
    | Home
    | SignUp
    | Login
    | File
    | DynamicForm
    | FormList
    | Composability of ComposabilityRoute
    | ValidationStrategies
    | CustomAction
    | NotFound


let private toHash page =
    let segmentsPart =
        match page with
        | Route.Home -> "home"
        | Route.SignUp -> "sign-up"
        | Route.Login -> "login"
        | Route.File -> "file"
        | Route.NotFound -> "not-found"
        | Route.DynamicForm -> "dynamic-form"
        | Route.Composability subRoute -> "composability/" + ComposabilityRoute.toSegment subRoute
        | Route.ValidationStrategies -> "validation-strategies"
        | Route.CustomAction -> "custom-actions"
        | Route.FormList -> "form-list"

    "#" + segmentsPart

let routeParser: Parser<Route->Route,Route> =
    oneOf
        [
            map Route.Home (s "home")
            map Route.SignUp (s "sign-up")
            map Route.Login (s "login")
            map Route.File (s "file")
            map Route.DynamicForm (s "dynamic-form")
            map Route.FormList (s "form-list")
            map Route.ValidationStrategies (s "validation-strategies")
            map Route.CustomAction (s "custom-actions")
            map (Route.Composability ComposabilityRoute.Simple) (s "composability" </> s "simple")
            map (Route.Composability ComposabilityRoute.WithConfiguration) (s "composability" </> s "with-configuration")
            map Route.Home top
        ]

let href route =
    prop.href (toHash route)

let modifyUrl route =
    route
    |> toHash
    |> Navigation.modifyUrl

let newUrl route =
    route
    |> toHash
    |> Navigation.newUrl

let modifyLocation route =
    window.location.href <- toHash route

File: examples/fableBuild/src/Router.js
import { Union } from "../.fable/fable-library.3.1.11/Types.js";
import { union_type } from "../.fable/fable-library.3.1.11/Reflection.js";
import { oneOf, top, s, map } from "../.fable/Fable.Elmish.Browser.3.0.4/parser.fs.js";
import { ofArray } from "../.fable/fable-library.3.1.11/List.js";
import { Navigation_newUrl, Navigation_modifyUrl } from "../.fable/Fable.Elmish.Browser.3.0.4/navigation.fs.js";

export class Route extends Union {
    constructor(tag, ...fields) {
        super();
        this.tag = (tag | 0);
        this.fields = fields;
    }
    cases() {
        return ["Home", "SignUp", "Login", "NotFound"];
    }
}

export function Route$reflection() {
    return union_type("Router.Route", [], Route, () => [[], [], [], []]);
}

function toHash(page) {
    const segmentsPart = (page.tag === 1) ? "sign-up" : ((page.tag === 2) ? "login" : ((page.tag === 3) ? "not-found" : "home"));
    return "#" + segmentsPart;
}

export const routeParser = (() => {
    const parsers = ofArray([map(new Route(0), s("home")), map(new Route(1), s("sign-up")), map(new Route(2), s("login")), map(new Route(0), (state) => top(state))]);
    return (state_1) => oneOf(parsers, state_1);
})();

export function href(route) {
    return ["href", toHash(route)];
}

export function modifyUrl(route) {
    return Navigation_modifyUrl(toHash(route));
}

export function newUrl(route) {
    return Navigation_newUrl(toHash(route));
}

export function modifyLocation(route) {
    window.location.href = toHash(route);
}

